Content.Server/EntityEffects/Effects/ActivateRandom.cs

```diff
@@ -0,0 +1,166 @@
+using Content.Shared.Administration.Logs;
+using Content.Shared.Chemistry.Reagent;
+using Content.Shared.Database;
+using Content.Shared.FixedPoint;
+using Content.Shared.EntityEffects;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+using JetBrains.Annotations;
+using System.Linq;
+
+namespace Content.Shared.EntityEffects.Effects;
+
+/// <summary>
+/// "Activates" random of predefined sets of chemicals, removing them from the solution
+/// and firing specified ReagentEffects once.
+/// </summary>
+[UsedImplicitly]
+public sealed partial class ActivateRandom : EntityEffect {
+    /// <summary>
+    /// List of sets of activatable chemicals along with results of the activation.
+    /// </summary>
+    [DataField(required: true)]
+    public ActivateRandomCase[] Cases;
+
+    protected override string? ReagentEffectGuidebookText(IPrototypeManager prototype, IEntitySystemManager entSys)
+    {
+        return Loc.GetString("reagent-effect-guidebook-activate-random",
+            ("chance", Probability),
+            ("cases", string.Join("\n", Cases.ToList().Select(case_ =>
+                "- "
+                + string.Join(", ", case_.Activates.ToList()
+                    .Select(activate =>
+                        prototype.TryIndex(activate.Reagent.Prototype, out ReagentPrototype? reagentProto)
+                            ? Loc.GetString("reagent-effect-guidebook-activate-random-activate",
+                                ("factor", activate.Factor),
+                                ("catalyst", !activate.Consume),
+                                ("reagent", reagentProto.LocalizedName))
+                            : null)
+                        .Where(x => x is not null))
+                + ":\n"
+                + string.Join("\n", case_.Effects.ToList()
+                    .Select(effect =>
+                    {
+                        var desc = effect.GuidebookEffectDescription(prototype, entSys);
+                        return desc is null ? null : "  " + desc;
+                    })
+                    .Where(x => x is not null))))
+                + "\n"));
+    }
```

Comment:
LINQ allocates for every chain in C# and doing this by hand will probably make this a lot more readable.